pkg/api/api.go

```diff
@@ -0,0 +1,322 @@
+	"github.com/cortexproject/cortex/pkg/distributor"
+	"github.com/cortexproject/cortex/pkg/ingester"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/querier"
+	"github.com/cortexproject/cortex/pkg/querier/frontend"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ruler"
+	"github.com/cortexproject/cortex/pkg/storegateway"
+	"github.com/cortexproject/cortex/pkg/util/push"
+)
+
+type Config struct {
+	AlertmanagerHTTPPrefix string `yaml:"alertmanager_http_prefix"`
+	PrometheusHTTPPrefix   string `yaml:"prometheus_http_prefix"`
+
+	// The following configs are injected by the upstream caller.
+	ServerPrefix       string          `yaml:"-"`
+	LegacyHTTPPrefix   string          `yaml:"-"`
+	HTTPAuthMiddleware middleware.Func `yaml:"-"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet with the set prefix.
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.StringVar(&cfg.AlertmanagerHTTPPrefix, prefix+"http.alertmanager-http-prefix", "/alertmanager", "HTTP URL path under which the Alertmanager ui and api will be served.")
+	f.StringVar(&cfg.PrometheusHTTPPrefix, prefix+"http.prometheus-http-prefix", "/prometheus", "TTP URL path under which the Prometheus api will be served.")
+}
+
+type API struct {
+	cfg            Config
+	authMiddleware middleware.Func
+	server         *server.Server
+	logger         log.Logger
+}
+
+func New(cfg Config, s *server.Server, logger log.Logger) (*API, error) {
+	// Ensure the encoded path is used. Required for the rules API
+	s.HTTP.UseEncodedPath()
+
+	api := &API{
+		cfg:            cfg,
+		authMiddleware: cfg.HTTPAuthMiddleware,
+		server:         s,
+		logger:         logger,
+	}
+
+	// If no authentication middleware is present in the config, use the default authentication middleware.
+	if cfg.HTTPAuthMiddleware == nil {
+		api.authMiddleware = middleware.AuthenticateUser
+	}
+
+	return api, nil
+}
+
+func (a *API) registerRoute(path string, handler http.Handler, auth bool, methods ...string) {
+	level.Debug(a.logger).Log("msg", "api: registering route", "methods", strings.Join(methods, ","), "path", path, "auth", auth)
+	if auth {
+		handler = a.authMiddleware.Wrap(handler)
+	}
+	if len(methods) == 0 {
+		a.server.HTTP.Path(path).Handler(handler)
+		return
+	}
+	a.server.HTTP.Path(path).Methods(methods...).Handler(handler)
+}
+
+func (a *API) registerRoutesWithPrefix(path string, handler http.Handler, auth bool, methods ...string) {
+	level.Debug(a.logger).Log("msg", "api: registering route", "methods", strings.Join(methods, ","), "path", path, "auth", auth)
+	if auth {
+		handler = a.authMiddleware.Wrap(handler)
+	}
+	if len(methods) == 0 {
+		a.server.HTTP.Path(path).Handler(handler)
```

Comment:
Prefix?

Reply:
Sorry about that, I was just looking into the failed integration tests and that looked to be the issue. Also the primary reason for not explicitly registering the alertmanager routes is because the Alertmanager UI routing is relatively complex and I didn't want to handle it in this PR. I also think it is fine to serve the Alertmanager directly if it is routed under the `/alertmanager/` prefix.